src/jdk.jlink/share/classes/jdk/tools/jlink/internal/ImageFileCreator.java

```diff
@@ -262,14 +263,15 @@ public String getString(int id) {
                 return writer.getString(id);
             }
         });
-
-        for (Archive archive : archives) {
-            String mn = archive.moduleName();
-            entriesForModule.get(mn).stream()
-                .map(e -> new ArchiveEntryResourcePoolEntry(mn,
+        archives.stream()
+                .sorted(Comparator.comparing(Archive::moduleName))
+                .forEach((archive) -> {
+                    String mn = archive.moduleName();
+                    entriesForModule.get(mn).stream()
+                            .map(e -> new ArchiveEntryResourcePoolEntry(mn,
                                     e.getResourcePoolEntryName(), e))
-                .forEach(resources::add);
-        }
+                            .forEach(resources::add);
+                });
```

Comment:
I'm somewhat allergic to nested forEach plus statement lambdas. I note that the original code mixed an imperative for-loop with a stream, which was kind of odd; converting to a straight nested for-loop might be reasonable. Alternatively, the nested forEaches could be converted to a flatMap as follows:

`
        archives.stream()
                .map(Archive::moduleName)
                .sorted()
                .flatMap(mn -> entriesForModule.get(mn).stream()
                               .map(e -> new ArchiveEntryResourcePoolEntry(mn, e.getResourcePoolEntryName(), e)))
                .forEach(resources::add);
`

Reply:
This is a good catch. I'm partial to flatMap, but unsure if that gets too clever what we're trying to do here or not.